### PR TITLE
Fix differences in the output of native and jgit

### PR DESCRIPTION
**Pull request by hgschmie**
On my box (using git 1.9.3), for a small project, there are the
following differences between jgit and native output:

```
field                   | jgit                       | native
git.build.user.name     | Henning Schmiedehausen     | "Henning Schmiedehausen"
git.build.user.email    | henning@schmiedehausen.org | "henning@schmiedehausen.org"
git.dirty               | true                       | false
git.commit.user.name    | Henning Schmiedehausen     | "Henning Schmiedehausen"
git.commit.user.email   | henning@schmiedehausen.org | "henning@schmiedehausen.org"
git.commit.message.full | Create README.md           | "Create README.md"
git.commit.message.short| Create README.md           | "Create README.md"
git.commit.time         | 2014-11-21T18:47:34-0800   | "2014-11-21 18:47:34 -0800"
git.tags                | bar,foo                    | foo,bar,origin/master,origin/
```

This uses the following custom settings:

```xml
<configuration>
    <dateFormat>yyyy-MM-dd'T'HH:mm:ssZZ</dateFormat>
    <abbrevLength>10</abbrevLength>
    <gitDescribe>
        <always>true</always>
        <abbrev>7</abbrev>
        <dirty>-dirty</dirty>
        <forceLongFormat>false</forceLongFormat>
    </gitDescribe>
</configuration>
```

Especially the "git.dirty" and "git.tags" is catastrophic but all the
other fields also make it hard to parse *any* information across
different settings of the git provider.

This patch fixes all the differences. Now the output of jgit and
native are identical.

Not a single unit test broke with these output changes. This is
worrisome because it implies that test coverage at least for the
native provider is not good.

### CODE REVIEW

**Comment by ktoso** [comment]
Thanks for the PR, adding a test to cover exactly what you found would be really good (as part of this PR). Would you mind giving it a shot? I'll look in depth at this one on the train today - definitely something I'll want to pull in.

**Comment by hgschmie** [comment]
Are there any examples on how to do native integration tests? They seem to require git installed and therefore inherently brittle. (also I was pretty busy the last few weeks so I did not find time to work on this. Maybe this weekend).

**Comment by ktoso** [comment]
Tests are run using both jgit and native git - https://github.com/ktoso/maven-git-commit-id-plugin/blob/master/src/test/java/pl/project13/maven/git/GitCommitIdMojoIntegrationTest.java#L59
Yes, it's brittle and I do not recommend anyone to use native git unless they really have a good reason to do so (had people with super long builds where native helped).

Hope this helps!

**Comment by TheSnoozer** [comment]
I was one of the folks who refactored/implemented the native git ;)
I know that the native implemntation mighty be a bit brittle but it really improves the build performance!
The benchmark results can be obtained from https://github.com/ktoso/maven-git-commit-id-plugin/pull/118

Furthermore there is a problem with a shallow git clone when using the jgit implementation:
https://github.com/ktoso/maven-git-commit-id-plugin/issues/61
https://github.com/ktoso/maven-git-commit-id-plugin/issues/96

It might not be the best if the output differs but that's how software development works...so far I have never seen a bug free software/implementation ;)
If you need some help with the testcases just let me know...

**Comment by hgschmie** [comment]
added integration testing

- caught a problem with the BuildUser/BuildEmail and CommitUser/CommitEmail settings
- caught a real problem when setting up a sandbox where the template has empty directories.

Please verify, merge and release. :-)

**Comment by hgschmie** [comment]
Yah, build is green. Ready to merge. :-)

File: src/main/java/pl/project13/maven/git/JGitProvider.java [hunk 1]
```diff
@@ -66,13 +69,13 @@ protected void init() throws MojoExecutionException {
   @Override
   protected String getBuildAuthorName() {
     String userName = git.getConfig().getString("user", null, "name");
-    return userName;
+    return Objects.firstNonNull(userName, "");
```
Comment (ktoso) [inline src/main/java/pl/project13/maven/git/JGitProvider.java]: good one

**Comment by hgschmie** [comment]
Do you have an ETA here? I am pretty eager to update my builds to use that fixed plugin.

**Comment by ktoso** [comment]
If  you're happy with running on a snapshot for a week I'll release even right now.
Having an crazy week where not sleeping is not able to add enough hours which I'd need to this week :<

**Comment by ktoso** [comment]
Went through changes and looks good, by judgement of my sleepy brain :-)
Merged and I'll release a snapshot you can run with for a little bit until I bump the proper version

**Comment by hgschmie** [comment]
Thanks Konrad, I really appreciate it.

**Comment by ktoso** [comment]
@hgschmie, snapshot deployed to sonatype snapshots repo ( https://oss.sonatype.org/content/repositories/snapshots/pl/project13/maven/git-commit-id-plugin/2.1.13-SNAPSHOT/ ).

Thanks a lot for the PR! I'll try to get around to a proper release soon. Cheers!
